File: week-7&8/problem-set/prob8.py
"""Prob8 example.

This file belongs to the week 7 and 8 tree and hashing practice solutions. It uses ChainingHashTable, is_distinct to illustrate the main idea with a small runnable example.
"""

class ChainingHashTable:
    """Simple hash table that resolves collisions with chaining."""
    def __init__(self, size=100):
        """Run the   init   step used in this example.

        Args:
            size (int): The size input used by this function.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.size = size
        self.table = [[] for _ in range(size)]

    def hash_function(self, value):
        """Map a value to one bucket in the hash table.

        Args:
            value (int): The value input used by this function.

        Returns:
            int: The integer result produced by the algorithm.
        """
        return value % 2

    def insert(self, value):
        """Insert a new value into the hash table if it is not already present.

        Args:
            value (int): The value input used by this function.

        Returns:
            object: The value produced by this function.
        """
        index = self.hash_function(value)
        print(index)
        if value not in self.table[index]:
            self.table[index].append(value)
            return True
        else:
            return False

def is_distinct(lst):
    """Return True when the list contains no duplicate values.

    Args:
        lst (object): The lst input used by this function.

    Returns:
        bool: True when the required condition is satisfied, otherwise False.
    """
    hash_table = ChainingHashTable()
    for elem in lst:
        if not hash_table.insert(elem):
            return False
    return True

# Example usage
lst = [1, 11, 21, 31]
print(is_distinct(lst))  # Output: True

File: week-7&8/problem-set/prob8.cpp
/**
 * @file prob8.cpp
 * @brief C++ version of the chaining-hash-table distinctness example.
 */

#include <iostream>
#include <vector>

using namespace std;

class ChainingHashTable {
public:
    /**
     * @brief Create a simple chaining hash table.
     * @param size int Number of buckets to create. The bucket array stores vectors of int.
     */
    explicit ChainingHashTable(int size = 100) : size(size), table(size) {}

    /**
     * @brief Map a value to one bucket in the hash table.
     * @param value int Value to hash.
     * @return int Bucket index used for value.
     */
    int hash_function(int value) const {
        return value % 2;
    }

    /**
     * @brief Insert a value if it is not already present.
     * @param value int Value to insert.
     * @return bool True when the value was inserted, otherwise false when it was already present.
     */
    bool insert(int value) {
        int index = hash_function(value);
        cout << index << '\n';
        for (int stored : table[index]) {
            if (stored == value) {
                return false;
            }
        }
        table[index].push_back(value);
        return true;
    }

private:
    int size;
    vector<vector<int>> table;
};

/**
 * @brief Return true when a list contains no duplicate values.
 * @param lst const vector<int>& List of integers to test.
 * @return bool True when all values are distinct, otherwise false.
 */
bool is_distinct(const vector<int>& lst) {
    ChainingHashTable hash_table;
    for (int elem : lst) {
        if (!hash_table.insert(elem)) {
            return false;
        }
    }
    return true;
}

/**
 * @brief Run the distinctness demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> lst = {1, 11, 21, 31};
    cout << boolalpha << is_distinct(lst) << '\n';
    return 0;
}
